server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -122,56 +120,62 @@ public CompletionStage<Void> restore(ZipFile zip) {
+            log.debugf("Restoring Cache %s: %s", cacheName, config.toStringConfiguration(cacheName));
+            // Create the cache
+            SecurityActions.getOrCreateCache(cm, cacheName, config);
+         } catch (IOException e) {
+            throw new CacheException(e);
+         }
 
-      AdvancedCache<Object, Object> cache = cm.getCache(cacheName).getAdvancedCache();
-      ComponentRegistry cr = SecurityActions.getCacheComponentRegistry(cache);
-      CommandsFactory commandsFactory = cr.getCommandsFactory();
-      KeyPartitioner keyPartitioner = cr.getComponent(KeyPartitioner.class);
-      InvocationHelper invocationHelper = cr.getComponent(InvocationHelper.class);
-      StorageConfigurationManager scm = cr.getComponent(StorageConfigurationManager.class);
-      PersistenceMarshaller persistenceMarshaller = cr.getPersistenceMarshaller();
-      Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
+         // Process .dat
+         String dataFile = dataFile(cacheName);
+         String data = cacheRoot.resolve(dataFile).toString();
+         ZipEntry zipEntry = zip.getEntry(data);
+         if (zipEntry == null)
+            return null;
 
-      boolean keyMarshalling = !scm.getKeyStorageMediaType().isBinary();
-      boolean valueMarshalling = !scm.getValueStorageMediaType().isBinary();
+         return new DataInputStream(zip.getInputStream(zipEntry));
+      }, is -> {
+         if (is == null)
+            return Flowable.empty();
 
-      SerializationContextRegistry ctxRegistry = SecurityActions.getGlobalComponentRegistry(cm).getComponent(SerializationContextRegistry.class);
-      ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+         AdvancedCache<Object, Object> cache = cm.getCache(cacheName).getAdvancedCache();
+         ComponentRegistry cr = SecurityActions.getCacheComponentRegistry(cache);
+         CommandsFactory commandsFactory = cr.getCommandsFactory();
+         KeyPartitioner keyPartitioner = cr.getComponent(KeyPartitioner.class);
+         InvocationHelper invocationHelper = cr.getComponent(InvocationHelper.class);
+         StorageConfigurationManager scm = cr.getComponent(StorageConfigurationManager.class);
+         PersistenceMarshaller persistenceMarshaller = cr.getPersistenceMarshaller();
+         Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
 
-      try {
-         DataInputStream is = new DataInputStream(zip.getInputStream(zipEntry));
+         boolean keyMarshalling = !scm.getKeyStorageMediaType().isBinary();
+         boolean valueMarshalling = !scm.getValueStorageMediaType().isBinary();
+
+         SerializationContextRegistry ctxRegistry = SecurityActions.getGlobalComponentRegistry(cm).getComponent(SerializationContextRegistry.class);
+         ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+
+         int batchSize = (int) (Long.highestOneBit(ProcessorInfo.availableProcessors()) << 1);
```

Comment:
You might want to increase this batch size a bit, this isn't a parallelized operation in regards to the CPU. It will only be parallel when the put command is asynchronous and that is done via network most likely by updating backups or through a store async.

Reply:
I would suggest using the state transfer chunk size to allow tuning.